testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jpa/appclient/PersistenceUnitInAppClientArchiveInServerTestCase.java

```diff
@@ -0,0 +1,80 @@
+import org.jboss.shrinkwrap.api.spec.EnterpriseArchive;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * Cover testing of https://issues.redhat.com/browse/WFLY-20277 change to ignore persistence units in app-client container archive when deploying on server.
+ *
+ * This test includes a client container archive that has a persistence.xml with invalid persistence provider name that would produce a deployment failure
+ * if the the persistence unit was actually started (since the invalid persistence provider class does not exist).
+ *
+ * @author Scott Marlow
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+public class PersistenceUnitInAppClientArchiveInServerTestCase {
+
+    @ArquillianResource
+    private Deployer deployer;
+
+    private static final String ARCHIVE_NAME = "PersistenceUnitInAppClientArchiveInServerTestCase";
+    @Deployment(name = ARCHIVE_NAME, managed = false)
+    public static Archive<?> deploy() {
+        EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, ARCHIVE_NAME + ".ear");
+
+        JavaArchive clientModule = ShrinkWrap.create(JavaArchive.class,"appclientcontainerarchive.jar");
+        clientModule.addAsManifestResource( PersistenceUnitInAppClientArchiveInServerTestCase.class.getPackage(), "application-client.xml","application-client.xml");
+        clientModule.addAsManifestResource(PersistenceUnitInAppClientArchiveInServerTestCase.class.getPackage(), "persistence.xml", "persistence.xml");
+        ear.addAsModule(clientModule);
+        JavaArchive jar = ShrinkWrap.create(JavaArchive.class, "jarfile.jar");
+        jar.addClass(MainArchiveEntity.class);
+        ear.addAsLibrary(jar);
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "Test.war");
+        war.addClasses(HttpRequest.class, SimpleServlet.class);
+        // WEB-INF/classes is implied
+        war.addAsWebInfResource(PersistenceUnitInAppClientArchiveInServerTestCase.class.getPackage(), "web.xml", "web.xml");
+        ear.addAsModule(war);
+
+        return ear;
+    }
+
+    /**
+     * Any deployment failures for this test means the fix for WFLY-20277 has been broken.
+     * This test is expected to simply pass meaning that the test deployment succeeded with the ignored invalid persistence provider.
+     */
+    @Test
+    public void testNothing() throws NamingException {
+        try {
+            deployer.deploy(ARCHIVE_NAME);
+        } catch (RuntimeException deploymentProblem) {
+            deploymentProblem.printStackTrace();
+            fail("The change for WFLY-20277 has been broken likely because the appclientcontainerarchive.jar contains an invalid persistence.xml (someone needs to verify that).  " +
+                    "Deployment failure message: " + deploymentProblem.getMessage());
+        }
+        finally {
+            deployer.undeploy(ARCHIVE_NAME);
```

Comment:
I think this should be in the try block right after the deploy.

If the deploy call fails, this undeploy will fail in the finally block (because it's not deployed) and that failure will hide the real test failure. It's only if the deploy succeeds that undeploy is needed.

Reply:
Sorry; probably better is to leave it after the try/catch but not in a finally. If it's in the try block an undeploy failure can't be distinguished from a deploy failure.